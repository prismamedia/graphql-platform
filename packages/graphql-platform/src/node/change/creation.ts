import * as utils from '@prismamedia/graphql-platform-utils';
import type { Node, NodeValue } from '../../node.js';
import { AbstractNodeChange } from '../abstract-change.js';

export class NodeCreation<
  TRequestContext extends object = any,
> extends AbstractNodeChange<TRequestContext> {
  public override readonly kind = utils.MutationType.CREATION;

  public readonly oldValue: undefined;
  public readonly newValue: Readonly<NodeValue>;

  public constructor(
    node: Node<TRequestContext>,
    requestContext: TRequestContext,
    maybeNewValue: unknown,
    createdAt?: Date,
    committedAt?: Date,
  ) {
    const newValue = Object.freeze(node.parseValue(maybeNewValue));

    super(
      node,
      node.mainIdentifier.parseValue(newValue),
      requestContext,
      createdAt,
      committedAt,
    );

    this.newValue = newValue;
  }
}
