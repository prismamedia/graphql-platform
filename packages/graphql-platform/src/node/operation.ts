import * as graphql from 'graphql';
import type { Constructor } from 'type-fest';
import type { Node } from '../node.js';
import { mutationConstructors, type Mutation } from './operation/mutation.js';
import { queryConstructors, type Query } from './operation/query.js';
import {
  subscriptionConstructors,
  type Subscription,
} from './operation/subscription.js';

export {
  argsPathKey,
  type NodeSelectionAwareArgs,
  type RawNodeSelectionAwareArgs,
} from './abstract-operation.js';
export * from './operation/api.js';
export * from './operation/context.js';
export * from './operation/custom.js';
export * from './operation/error.js';
export * from './operation/mutation.js';
export * from './operation/query.js';
export * from './operation/subscription.js';

export type OperationByType<TRequestContext extends object = any> = {
  [graphql.OperationTypeNode.MUTATION]: Mutation<TRequestContext>;
  [graphql.OperationTypeNode.QUERY]: Query<TRequestContext>;
  [graphql.OperationTypeNode.SUBSCRIPTION]: Subscription<TRequestContext>;
};

export type OperationType = keyof OperationByType;

export type Operation<TRequestContext extends object = any> =
  OperationByType<TRequestContext>[OperationType];

export const operationConstructorsByType = {
  [graphql.OperationTypeNode.MUTATION]: mutationConstructors,
  [graphql.OperationTypeNode.QUERY]: queryConstructors,
  [graphql.OperationTypeNode.SUBSCRIPTION]: subscriptionConstructors,
} as const satisfies {
  [TType in OperationType]: Constructor<OperationByType[TType], [Node]>[];
};

export type OperationsByType<TRequestContext extends object = any> = {
  [TType in OperationType]: OperationByType<TRequestContext>[TType][];
};
