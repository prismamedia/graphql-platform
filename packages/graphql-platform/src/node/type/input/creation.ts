import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/memoize';
import inflection from 'inflection';
import assert from 'node:assert/strict';
import type { Except } from 'type-fest';
import type { Edge, Node, NodeValue, ReverseEdge } from '../../../node.js';
import type { MutationContext } from '../../operation.js';
import type { NodeCreationValue } from '../../statement/creation.js';
import {
  ComponentCreationInput,
  FieldCreationInput,
  LeafCreationInput,
  ReverseEdgeCreationInput,
} from './creation/field.js';

export * from './creation/field.js';

export type NodeCreationInputValue = utils.Nillable<utils.PlainObject>;

export class NodeCreationInputType extends utils.ObjectInputType<FieldCreationInput> {
  readonly #excludedEdge?: Edge;

  public constructor(
    public readonly node: Node,
    excludedEdge?: Edge,
  ) {
    excludedEdge && node.ensureEdge(excludedEdge);
    assert(node.isCreatable(), `The "${node}" node is not creatable`);

    super({
      name: [
        node.name,
        inflection.camelize(utils.MutationType.CREATION),
        excludedEdge
          ? `Without${inflection.camelize(excludedEdge.name)}`
          : undefined,
        'Input',
      ]
        .filter(Boolean)
        .join(''),
      description: `The "${node}" node's ${utils.MutationType.CREATION}`,
    });

    this.#excludedEdge = excludedEdge;
  }

  @Memoize()
  protected get componentFields(): ReadonlyArray<ComponentCreationInput> {
    return Array.from(this.node.componentsByName.values()).reduce<
      ComponentCreationInput[]
    >((fields, component) => {
      if (component !== this.#excludedEdge) {
        fields.push(component.creationInput);
      }

      return fields;
    }, []);
  }

  @Memoize()
  protected get reverseEdgeFields(): ReadonlyArray<ReverseEdgeCreationInput> {
    return Array.from(this.node.reverseEdgesByName.values()).reduce<
      ReverseEdgeCreationInput[]
    >((fields, reverseEdge) => {
      if (reverseEdge.creationInput) {
        fields.push(reverseEdge.creationInput);
      }

      return fields;
    }, []);
  }

  @Memoize()
  protected get virtualFields(): ReadonlyArray<utils.Input> {
    return this.node.features.flatMap((feature) => {
      const { config, configPath } = feature.getMutationConfig(
        utils.MutationType.CREATION,
      );

      const virtualFieldsConfig = config?.virtualFields;
      const virtualFieldsConfigPath = utils.addPath(
        configPath,
        'virtualFields',
      );

      utils.assertNillablePlainObject(
        virtualFieldsConfig,
        virtualFieldsConfigPath,
      );

      return virtualFieldsConfig
        ? utils.aggregateGraphError<
            [utils.Name, Except<utils.InputConfig, 'name'>],
            utils.Input[]
          >(
            Object.entries(virtualFieldsConfig),
            (fields, [virtualFieldName, virtualFieldConfig]) => {
              if (
                this.componentFields.find(
                  (field) => field.name === virtualFieldName,
                )
              ) {
                throw new utils.UnexpectedValueError(
                  `not to have a component-field's name`,
                  virtualFieldName,
                  { path: virtualFieldsConfigPath },
                );
              } else if (
                this.reverseEdgeFields.find(
                  (field) => field.name === virtualFieldName,
                )
              ) {
                throw new utils.UnexpectedValueError(
                  `not to have a reverse-edge-field's name`,
                  virtualFieldName,
                  { path: virtualFieldsConfigPath },
                );
              }

              return [
                ...fields,
                new utils.Input(
                  { ...virtualFieldConfig, name: virtualFieldName },
                  utils.addPath(virtualFieldsConfigPath, virtualFieldName),
                ),
              ];
            },
            [],
            { path: virtualFieldsConfigPath },
          )
        : [];
    });
  }

  @Memoize()
  public override get fields(): ReadonlyArray<FieldCreationInput> {
    return [
      ...this.componentFields,
      ...this.reverseEdgeFields,
      ...this.virtualFields,
    ];
  }

  @Memoize()
  public override isPublic(): boolean {
    return (
      this.node.isPubliclyMutable(utils.MutationType.CREATION) &&
      super.isPublic() &&
      this.componentFields.some((field) => field.isPublic())
    );
  }

  public async resolveValue(
    data: Readonly<NonNullable<NodeCreationInputValue>>,
    context: MutationContext,
    path?: utils.Path,
  ): Promise<NodeCreationValue> {
    const resolvedValue: NodeCreationValue = Object.create(null);

    for (const field of this.componentFields) {
      const fieldData = data[field.name];

      const componentValue =
        fieldData == null || field instanceof LeafCreationInput
          ? fieldData
          : await field.resolveValue(
              fieldData,
              context,
              utils.addPath(path, field.name),
            );

      if (componentValue !== undefined) {
        Object.assign(resolvedValue, { [field.name]: componentValue });
      }
    }

    return resolvedValue;
  }

  public hasReverseEdgeActions(
    data: Readonly<NonNullable<NodeCreationInputValue>>,
    reverseEdges: ReadonlyArray<ReverseEdge> = Array.from(
      this.node.reverseEdgeSet,
    ),
  ): boolean {
    return this.reverseEdgeFields.some(
      (field) =>
        data[field.name] != null &&
        field.hasActions(data[field.name]) &&
        reverseEdges.includes(field.reverseEdge),
    );
  }

  public async applyReverseEdgeActions(
    nodeValue: Readonly<NodeValue>,
    data: Readonly<NonNullable<NodeCreationInputValue>>,
    context: MutationContext,
    path?: utils.Path,
  ): Promise<void> {
    for (const field of this.reverseEdgeFields) {
      const fieldData = data[field.name];

      if (fieldData != null && field.hasActions(fieldData)) {
        await field.applyActions(
          nodeValue,
          fieldData,
          context,
          utils.addPath(path, field.name),
        );
      }
    }
  }
}
