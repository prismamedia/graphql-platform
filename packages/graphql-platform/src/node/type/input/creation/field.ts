import type { Input } from '@prismamedia/graphql-platform-utils';
import type { ComponentCreationInput } from './field/component.js';
import type { ReverseEdgeCreationInput } from './field/reverse-edge.js';

export * from './field/component.js';
export * from './field/reverse-edge.js';

export type FieldCreationInput =
  | ComponentCreationInput
  | ReverseEdgeCreationInput
  | Input;
