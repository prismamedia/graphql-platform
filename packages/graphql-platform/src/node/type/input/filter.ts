import {
  jsonScalarTypes,
  Scalars,
  stringScalarTypesByName,
} from '@prismamedia/graphql-platform-scalars';
import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/ts-memoize';
import * as graphql from 'graphql';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { Edge, Node } from '../../../node.js';
import { Leaf, type LeafValue } from '../../definition/component/leaf.js';
import type { ReverseEdgeMultiple } from '../../definition/reverse-edge/multiple.js';
import { ReverseEdgeUnique } from '../../definition/reverse-edge/unique.js';
import type { OperationContext } from '../../operation/context.js';
import {
  AndOperation,
  BooleanFilter,
  BooleanValue,
  EdgeExistsFilter,
  LeafComparisonFilter,
  LeafFullTextFilter,
  LeafInFilter,
  NodeFilter,
  NotOperation,
  OrOperation,
  ReverseEdgeMultipleCountFilter,
  ReverseEdgeMultipleExistsFilter,
  ReverseEdgeUniqueExistsFilter,
} from '../../statement/filter.js';
import {
  BooleanOperationFilterInputType,
  EdgeFilterInputType,
  FieldFilterInputType,
  LeafFilterInputType,
  ReverseEdgeFilterInputType,
} from './filter/field.js';

export * from './filter/field.js';

export type NodeFilterInputValue = utils.Nillable<utils.PlainObject>;

export class NodeFilterInputType extends utils.ObjectInputType<FieldFilterInputType> {
  public constructor(public readonly node: Node) {
    super({
      name: `${node}FilterInput`,
      description: `The "${node}" nodes' filter`,
    });
  }

  protected getLeafFields(leaf: Leaf): LeafFilterInputType[] {
    const fields: LeafFilterInputType[] = [];

    // eq, not
    fields.push(
      new LeafFilterInputType<LeafValue>(leaf, 'eq', {
        type: utils.nonNullableInputTypeDecorator(
          leaf.type,
          !leaf.isNullable(),
        ),
        filter: (value, _context, _path) =>
          new LeafComparisonFilter(leaf, 'eq', value),
      }),
      new LeafFilterInputType<LeafValue>(leaf, 'not', {
        type: utils.nonNullableInputTypeDecorator(
          leaf.type,
          !leaf.isNullable(),
        ),
        filter: (value, _context, _path) =>
          new NotOperation(new LeafComparisonFilter(leaf, 'eq', value)),
      }),
    );

    // is_null
    if (leaf.isNullable()) {
      fields.push(
        new LeafFilterInputType<boolean>(leaf, 'is_null', {
          type: new utils.NonNullableInputType(Scalars.Boolean),
          filter: (value, _context, _path) =>
            value
              ? new LeafComparisonFilter(leaf, 'eq', null)
              : new NotOperation(new LeafComparisonFilter(leaf, 'eq', null)),
        }),
      );
    }

    // in, not_in
    if (
      graphql.isEnumType(leaf.type) ||
      (leaf.type === Scalars.Boolean && leaf.isNullable()) ||
      ![...jsonScalarTypes, Scalars.DraftJS].includes(leaf.type as any)
    ) {
      fields.push(
        new LeafFilterInputType<LeafValue[]>(leaf, 'in', {
          type: new utils.NonNullableInputType(
            new utils.ListableInputType(
              new utils.NonOptionalInputType(
                utils.nonNullableInputTypeDecorator(
                  leaf.type,
                  !leaf.isNullable(),
                ),
              ),
            ),
          ),
          filter: (values, _context, _path) => new LeafInFilter(leaf, values),
        }),
        new LeafFilterInputType<LeafValue[]>(leaf, 'not_in', {
          type: new utils.NonNullableInputType(
            new utils.ListableInputType(
              new utils.NonOptionalInputType(
                utils.nonNullableInputTypeDecorator(
                  leaf.type,
                  !leaf.isNullable(),
                ),
              ),
            ),
          ),
          filter: (values, _context, _path) =>
            new NotOperation(new LeafInFilter(leaf, values)),
        }),
      );
    }

    // gt, gte, lt, lte
    if (leaf.isSortable()) {
      for (const operator of ['gt', 'gte', 'lt', 'lte'] as const) {
        fields.push(
          new LeafFilterInputType<LeafValue>(leaf, operator, {
            type: new utils.NonNullableInputType(leaf.type),
            filter: (value, _context, _path) =>
              new LeafComparisonFilter(leaf, operator, value),
          }),
        );
      }
    }

    // Full-text search
    // contains, starts_with, ends_with
    if (
      [
        ...Object.values(stringScalarTypesByName),
        Scalars.DraftJS,
        Scalars.JSONObject,
      ].includes(leaf.type as any)
    ) {
      for (const operator of [
        'contains',
        'starts_with',
        'ends_with',
      ] as const) {
        fields.push(
          new LeafFilterInputType<string>(leaf, operator, {
            type: new utils.NonNullableInputType(Scalars.NonEmptyString),
            filter: (value, _context, _path) =>
              new LeafFullTextFilter(leaf, operator, value),
          }),
          new LeafFilterInputType<string>(leaf, `not_${operator}`, {
            type: new utils.NonNullableInputType(Scalars.NonEmptyString),
            filter: (value, _context, _path) =>
              new NotOperation(new LeafFullTextFilter(leaf, operator, value)),
          }),
        );
      }
    }

    return fields;
  }

  protected getEdgeFields(edge: Edge): EdgeFilterInputType[] {
    const fields: EdgeFilterInputType[] = [];

    fields.push(
      new EdgeFilterInputType<NodeFilterInputValue>(edge, 'eq', {
        type: utils.nonNullableInputTypeDecorator(
          edge.head.filterInputType,
          !edge.isNullable(),
        ),
        filter: (value, context, path) =>
          value === null
            ? new NotOperation(new EdgeExistsFilter(edge))
            : new EdgeExistsFilter(
                edge,
                edge.head.filterInputType.filter(value, context, path),
              ),
      }),
      new EdgeFilterInputType<NodeFilterInputValue>(edge, 'not', {
        type: utils.nonNullableInputTypeDecorator(
          edge.head.filterInputType,
          !edge.isNullable(),
        ),
        filter: (value, context, path) =>
          value === null
            ? new EdgeExistsFilter(edge)
            : new NotOperation(
                new EdgeExistsFilter(
                  edge,
                  edge.head.filterInputType.filter(value, context, path),
                ),
              ),
      }),
    );

    // is_null
    if (edge.isNullable()) {
      fields.push(
        new EdgeFilterInputType<boolean>(edge, 'is_null', {
          type: new utils.NonNullableInputType(Scalars.Boolean),
          filter: (value, _context, _path) =>
            value
              ? new NotOperation(new EdgeExistsFilter(edge))
              : new EdgeExistsFilter(edge),
        }),
      );
    }

    return fields;
  }

  protected getReverseEdgeUniqueFields(
    reverseEdge: ReverseEdgeUnique,
  ): ReverseEdgeFilterInputType[] {
    return [
      new ReverseEdgeFilterInputType<NodeFilterInputValue>(reverseEdge, 'eq', {
        type: reverseEdge.head.filterInputType,
        filter: (value, context, path) =>
          value === null
            ? new NotOperation(new ReverseEdgeUniqueExistsFilter(reverseEdge))
            : new ReverseEdgeUniqueExistsFilter(
                reverseEdge,
                reverseEdge.head.filterInputType.filter(value, context, path),
              ),
      }),
      new ReverseEdgeFilterInputType<NodeFilterInputValue>(reverseEdge, 'not', {
        type: reverseEdge.head.filterInputType,
        filter: (value, context, path) =>
          value === null
            ? new ReverseEdgeUniqueExistsFilter(reverseEdge)
            : new NotOperation(
                new ReverseEdgeUniqueExistsFilter(
                  reverseEdge,
                  reverseEdge.head.filterInputType.filter(value, context, path),
                ),
              ),
      }),
      new ReverseEdgeFilterInputType<NodeFilterInputValue>(
        reverseEdge,
        'is_null',
        {
          type: new utils.NonNullableInputType(Scalars.Boolean),
          filter: (value, _context, _path) =>
            value
              ? new NotOperation(new ReverseEdgeUniqueExistsFilter(reverseEdge))
              : new ReverseEdgeUniqueExistsFilter(reverseEdge),
        },
      ),
    ];
  }

  /**
   * We can note the following equivalences:
   *
   *  set.every(filter);
   *    = !set.some(!filter);
   *    = set.none(!filter);
   *
   *  set.some(filter);
   *    = !set.every(!filter);
   *    = !set.none(filter);
   *
   *  set.none(filter);
   *    = !set.some(filter);
   *    = set.every(!filter);
   */
  protected getReverseEdgeMultipleFields(
    reverseEdge: ReverseEdgeMultiple,
  ): ReverseEdgeFilterInputType[] {
    return [
      new ReverseEdgeFilterInputType<NonNullable<NodeFilterInputValue>>(
        reverseEdge,
        'every',
        {
          type: new utils.NonNullableInputType(
            reverseEdge.head.filterInputType,
          ),

          // set.every(filter) = !set.some(!filter);
          filter: (value, context, path) =>
            new NotOperation(
              new ReverseEdgeMultipleExistsFilter(
                reverseEdge,
                reverseEdge.head.filterInputType.filter(
                  value,
                  context,
                  path,
                ).complement,
              ),
            ),
        },
      ),
      new ReverseEdgeFilterInputType<NonNullable<NodeFilterInputValue>>(
        reverseEdge,
        'some',
        {
          type: new utils.NonNullableInputType(
            reverseEdge.head.filterInputType,
          ),
          filter: (value, context, path) =>
            new ReverseEdgeMultipleExistsFilter(
              reverseEdge,
              reverseEdge.head.filterInputType.filter(value, context, path),
            ),
        },
      ),
      new ReverseEdgeFilterInputType<NonNullable<NodeFilterInputValue>>(
        reverseEdge,
        'none',
        {
          type: new utils.NonNullableInputType(
            reverseEdge.head.filterInputType,
          ),

          // set.none(filter) = !set.some(filter);
          filter: (value, context, path) =>
            new NotOperation(
              new ReverseEdgeMultipleExistsFilter(
                reverseEdge,
                reverseEdge.head.filterInputType.filter(value, context, path),
              ),
            ),
        },
      ),
      ...(['eq', 'not', 'gt', 'gte', 'lt', 'lte'] as const).map(
        (operator) =>
          new ReverseEdgeFilterInputType<number>(reverseEdge, operator, {
            name:
              operator === 'eq'
                ? reverseEdge.countFieldName
                : `${reverseEdge.countFieldName}_${operator}`,
            type: new utils.NonNullableInputType(Scalars.UnsignedInt),
            filter: (value, _context, _path) =>
              operator === 'not'
                ? new NotOperation(
                    new ReverseEdgeMultipleCountFilter(
                      reverseEdge,
                      'eq',
                      value,
                    ),
                  )
                : new ReverseEdgeMultipleCountFilter(
                    reverseEdge,
                    operator,
                    value,
                  ),
          }),
      ),
    ];
  }

  protected getBooleanOperationFields(): BooleanOperationFilterInputType[] {
    return [
      new BooleanOperationFilterInputType<NodeFilterInputValue[]>({
        name: 'AND',
        type: new utils.NonNullableInputType(new utils.ListableInputType(this)),
        filter: (values, context, path) =>
          new AndOperation(
            utils.aggregateError<NodeFilterInputValue, BooleanFilter[]>(
              values,
              (operands, value, index) => [
                ...operands,
                this.filter(value, context, utils.addPath(path, index)).filter,
              ],
              [],
              { path },
            ),
          ),
      }),
      new BooleanOperationFilterInputType<NodeFilterInputValue[]>({
        name: 'OR',
        type: new utils.NonNullableInputType(new utils.ListableInputType(this)),
        filter: (values, context, path) =>
          new OrOperation(
            utils.aggregateError<NodeFilterInputValue, BooleanFilter[]>(
              values,
              (operands, value, index) => [
                ...operands,
                this.filter(value, context, utils.addPath(path, index)).filter,
              ],
              [],
              { path },
            ),
          ),
      }),
      new BooleanOperationFilterInputType<NodeFilterInputValue>({
        name: 'NOT',
        type: this,
        filter: (value, context, path) =>
          new NotOperation(this.filter(value, context, path).filter),
      }),
    ];
  }

  @Memoize()
  public override get fields(): ReadonlyArray<FieldFilterInputType> {
    return [
      ...Array.from(
        this.node.componentsByName.values(),
      ).flatMap<FieldFilterInputType>((component) =>
        component instanceof Leaf
          ? this.getLeafFields(component)
          : this.getEdgeFields(component),
      ),
      ...Array.from(
        this.node.reverseEdgesByName.values(),
      ).flatMap<FieldFilterInputType>((reverseEdge) =>
        reverseEdge instanceof ReverseEdgeUnique
          ? this.getReverseEdgeUniqueFields(reverseEdge)
          : this.getReverseEdgeMultipleFields(reverseEdge),
      ),
      ...this.getBooleanOperationFields(),
    ];
  }

  public filter<
    TRequestContext extends object,
    TConnector extends ConnectorInterface,
  >(
    value: NodeFilterInputValue,
    context?: OperationContext<TRequestContext, TConnector>,
    path: utils.Path = utils.addPath(undefined, this.name),
  ): NodeFilter<TRequestContext, TConnector> {
    return new NodeFilter(
      this.node,
      value === undefined
        ? new BooleanValue(true)
        : value === null
        ? new BooleanValue(false)
        : new AndOperation(
            Object.entries(value).map(([filterName, filterValue]) =>
              this.getField(filterName, path).filter(
                filterValue,
                context,
                utils.addPath(path, filterName),
              ),
            ),
          ),
    );
  }

  public parseAndFilter<
    TRequestContext extends object,
    TConnector extends ConnectorInterface,
  >(
    maybeValue: unknown,
    context?: OperationContext<TRequestContext, TConnector>,
    path: utils.Path = utils.addPath(undefined, this.name),
  ): NodeFilter<TRequestContext, TConnector> {
    return this.filter(this.parseValue(maybeValue, path), context, path);
  }
}
