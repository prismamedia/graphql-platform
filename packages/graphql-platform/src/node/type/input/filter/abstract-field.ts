import * as utils from '@prismamedia/graphql-platform-utils';
import type { Except } from 'type-fest';
import type { OperationContext } from '../../../operation/context.js';
import type { BooleanFilter } from '../../../statement/filter/boolean.js';

export type AbstractFieldFilterInputConfig<TValue> = Except<
  utils.InputConfig<TValue>,
  'defaultValue' | 'parser'
> & {
  filter(
    value: Exclude<TValue, undefined>,
    context: OperationContext | undefined,
    path: utils.Path,
  ): BooleanFilter;
};

export abstract class AbstractFieldFilterInput<
  TValue = any,
> extends utils.Input<TValue> {
  public readonly filter: AbstractFieldFilterInputConfig<TValue>['filter'];

  public constructor({
    filter,
    ...config
  }: AbstractFieldFilterInputConfig<TValue>) {
    super(config);

    this.filter = filter;
  }
}
