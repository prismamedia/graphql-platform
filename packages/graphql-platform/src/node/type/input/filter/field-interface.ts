import type * as utils from '@prismamedia/graphql-platform-utils';
import type { OperationContext } from '../../../operation.js';
import type { BooleanFilter } from '../../../statement.js';

export interface FieldFilterInputInterface<TValue = any>
  extends utils.Input<TValue> {
  filter(
    value: Exclude<TValue, undefined>,
    context: OperationContext | undefined,
    path: utils.Path,
  ): BooleanFilter;
}
