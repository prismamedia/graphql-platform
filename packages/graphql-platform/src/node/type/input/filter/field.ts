export * from './abstract-field.js';
export * from './field/boolean-operation.js';
export * from './field/edge.js';
export * from './field/leaf.js';
export * from './field/reverse-edge.js';
