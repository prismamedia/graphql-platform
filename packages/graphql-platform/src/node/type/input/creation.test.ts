import * as utils from '@prismamedia/graphql-platform-utils';
import * as graphql from 'graphql';
import { GraphQLPlatform, NodeCreationStatement } from '../../../index.js';
import {
  ArticleStatus,
  MyGP,
  myUserContext,
  nodeNames,
  nodes,
} from '../../../__tests__/config.js';
import { mockConnector } from '../../../__tests__/connector-mock.js';
import { MutationContext } from '../../operation.js';
import { LeafComparisonFilter } from '../../statement/filter.js';
import { NodeCreationInputType } from './creation.js';

describe('NodeCreationInputType', () => {
  let gp: MyGP;

  beforeAll(() => {
    gp = new GraphQLPlatform({
      nodes,
      connector: mockConnector({
        find: async ({ node, filter, selection }) => {
          if (
            node.name === 'Category' &&
            filter?.filter.equals(
              new LeafComparisonFilter(
                node.getLeafByName('id'),
                'eq',
                '91a7c846-b030-4ef3-aaaa-747fe7b11519',
              ),
            )
          ) {
            return [{ _id: 4 }];
          } else if (
            node.name === 'User' &&
            filter?.filter.equals(
              new LeafComparisonFilter(
                node.getLeafByName('id'),
                'eq',
                '2059b77a-a735-41fe-b415-5b12944b6ba6',
              ),
            )
          ) {
            return [
              (selection.expressionsByKey.has('id')
                ? { id: '2059b77a-a735-41fe-b415-5b12944b6ba6' }
                : { username: 'yvann' }) as any,
            ];
          }

          throw new Error('KO');
        },
      }),
    });
  });

  describe('Definition', () => {
    it.each(nodeNames)('%s has a creation input type', (nodeName) => {
      const node = gp.getNodeByName(nodeName);

      const creationInputType = node.creationInputType;
      expect(creationInputType).toBeInstanceOf(NodeCreationInputType);

      if (node.isMutationPublic(utils.MutationType.CREATION)) {
        expect(creationInputType.getGraphQLInputType()).toBeInstanceOf(
          graphql.GraphQLInputObjectType,
        );

        expect(
          graphql.printType(creationInputType.getGraphQLInputType()),
        ).toMatchSnapshot(creationInputType.name);
      } else {
        expect(() => creationInputType.getGraphQLInputType()).toThrowError(
          `The "${nodeName}CreationInput" input type is private`,
        );
      }
    });
  });

  describe('Runtime', () => {
    describe('Fails', () => {
      it.each([
        [
          'Article',
          {},
          '"ArticleCreationInput.title" - Expects a non-undefined "NonEmptyTrimmedString"',
        ],
        [
          'Article',
          {
            title: 'My valid title',
            category: {},
          },
          '"ArticleCreationInput.category" - Expects one and only one action, got: [Object: null prototype] {}',
        ],
        [
          'Article',
          {
            title: 'My valid title',
            category: {
              connect: { id: '91a7c846-b030-4ef3-aaaa-747fe7b11519' },
              connectIfExists: { id: '91a7c846-b030-4ef3-aaaa-747fe7b11519' },
            },
          },
          '"ArticleCreationInput.category" - Expects one and only one action, got: [Object: null prototype] {',
        ],
      ])(
        '%sCreationInput.parseValue(%p) throws an error',
        (nodeName, value, error) => {
          const node = gp.getNodeByName(nodeName);
          const creationInputType = node.creationInputType;

          expect(() => creationInputType.parseValue(value)).toThrowError(error);
        },
      );
    });

    describe('Works', () => {
      it('creates a valid statement', async () => {
        const Article = gp.getNodeByName('Article');
        const ArticleCreationInputType = Article.creationInputType;

        const input: utils.PlainObject = {
          title: "My article's title",
          status: undefined,
          category: {
            connectIfExists: { id: '91a7c846-b030-4ef3-aaaa-747fe7b11519' },
          },
          createdBy: {
            connect: { id: '2059b77a-a735-41fe-b415-5b12944b6ba6' },
          },
          updatedBy: {
            connect: { id: '2059b77a-a735-41fe-b415-5b12944b6ba6' },
          },
        };

        const parsedValue = ArticleCreationInputType.parseValue(input)!;

        expect(parsedValue).toEqual({
          id: expect.any(String),
          title: "My article's title",
          status: ArticleStatus.DRAFT,
          category: {
            connectIfExists: { id: '91a7c846-b030-4ef3-aaaa-747fe7b11519' },
          },
          createdBy: {
            connect: { id: '2059b77a-a735-41fe-b415-5b12944b6ba6' },
          },
          createdAt: expect.any(Date),
          updatedBy: {
            connect: { id: '2059b77a-a735-41fe-b415-5b12944b6ba6' },
          },
          updatedAt: expect.any(Date),
          views: 0n,
          score: 0.5,
        });

        const resolvedValue = await ArticleCreationInputType.resolveValue(
          parsedValue,
          new MutationContext(gp, myUserContext),
        );

        expect(resolvedValue).toEqual({
          id: expect.any(String),
          title: "My article's title",
          status: ArticleStatus.DRAFT,
          category: { _id: 4 },
          createdBy: { id: '2059b77a-a735-41fe-b415-5b12944b6ba6' },
          createdAt: expect.any(Date),
          updatedBy: { username: 'yvann' },
          updatedAt: expect.any(Date),
          views: 0n,
          score: 0.5,
        });

        const statement = new NodeCreationStatement(Article, resolvedValue);
        expect(statement).toBeInstanceOf(NodeCreationStatement);

        expect(statement.value).toEqual({
          id: expect.any(String),
          title: "My article's title",
          status: ArticleStatus.DRAFT,
          category: { _id: 4 },
          createdBy: { id: '2059b77a-a735-41fe-b415-5b12944b6ba6' },
          createdAt: expect.any(Date),
          updatedBy: { username: 'yvann' },
          updatedAt: expect.any(Date),
          views: 0n,
          score: 0.5,
        });
      });
    });
  });
});
