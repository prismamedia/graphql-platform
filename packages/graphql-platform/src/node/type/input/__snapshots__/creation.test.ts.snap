// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeCreationInputType Definition Article has a creation input type: ArticleCreationInput 1`] = `
""""The "Article" node's creation"""
input ArticleCreationInput {
  """You can either provide an UUID or let one be generated for you"""
  id: UUIDv4
  status: ArticleStatus! = DRAFT
  title: NonEmptyTrimmedString!

  """You can either provide a slug or let the title be "slugified" for you"""
  slug: NonEmptyTrimmedString

  """The article's body"""
  body: DraftJS
  category: ArticleNestedCategoryEdgeCreationInput

  """Contains any arbitrary data you want to store alongside the article"""
  metas: JSONObject

  """Is the article highlighted?"""
  highlighted: Boolean

  """Is the article a partnership?"""
  sponsored: Boolean
  tags: ArticleNestedTagsReverseEdgeCreationInput

  """It is possible to provide the article's body as raw HTML"""
  htmlBody: String
}"
`;

exports[`NodeCreationInputType Definition ArticleTag has a creation input type: ArticleTagCreationInput 1`] = `
""""The "ArticleTag" node's creation"""
input ArticleTagCreationInput {
  article: ArticleTagNestedArticleEdgeCreationInput!
  tag: ArticleTagNestedTagEdgeCreationInput!
  order: UnsignedInt!
  moderations: ArticleTagNestedModerationsReverseEdgeCreationInput
}"
`;

exports[`NodeCreationInputType Definition ArticleTagModeration has a creation input type: ArticleTagModerationCreationInput 1`] = `
""""The "ArticleTagModeration" node's creation"""
input ArticleTagModerationCreationInput {
  articleTag: ArticleTagModerationNestedArticleTagEdgeCreationInput!
  moderator: ArticleTagModerationNestedModeratorEdgeCreationInput!
  moderation: NonEmptyTrimmedString
}"
`;

exports[`NodeCreationInputType Definition Category has a creation input type: CategoryCreationInput 1`] = `
""""The "Category" node's creation"""
input CategoryCreationInput {
  """You can either provide an UUID or let one be generated for you"""
  id: UUIDv4
  title: NonEmptyTrimmedString!

  """You can either provide a slug or let the title be "slugified" for you"""
  slug: NonEmptyTrimmedString
  parent: CategoryNestedParentEdgeCreationInput
  order: UnsignedInt!

  """This category's children"""
  children: CategoryNestedChildrenReverseEdgeCreationInput

  """The articles attached to this category"""
  articles: CategoryNestedArticlesReverseEdgeCreationInput

  """It is possible to provide the article's body as raw HTML"""
  htmlBody: String
}"
`;

exports[`NodeCreationInputType Definition Tag has a creation input type: TagCreationInput 1`] = `
""""The "Tag" node's creation"""
input TagCreationInput {
  """You can either provide an UUID or let one be generated for you"""
  id: UUIDv4

  """A tag can be deprecated"""
  deprecated: Boolean
  title: NonEmptyTrimmedString!

  """You can either provide a slug or let the title be "slugified" for you"""
  slug: NonEmptyTrimmedString
  articles: TagNestedArticlesReverseEdgeCreationInput
}"
`;

exports[`NodeCreationInputType Definition UserProfile has a creation input type: UserProfileCreationInput 1`] = `
""""The "UserProfile" node's creation"""
input UserProfileCreationInput {
  user: UserProfileNestedUserEdgeCreationInput!
  birthday: Date
  facebookId: ID
  googleId: ID
  twitterHandle: ID
}"
`;
